src/core/qlist.cc

```diff
@@ -385,6 +386,25 @@ void QList::SetPackedThreshold(unsigned threshold) {
   packed_threshold = threshold;
 }
 
+size_t QList::DefragIfNeeded(PageUsage* page_usage) {
+  size_t reallocated = 0;
+  for (Node* curr = head_; curr; curr = curr->next) {
```

Comment:
did you consider reallocating the nodes themselves? 
btw, I do not think it is needed but at least add a comment:
`// we do not reallocate nodes because their size is constant so it is expected that their pages will usually be unfragmented` or smthing like this